File: examples/ruby-rspec/lib/library.rb
# frozen_string_literal: true

require 'date'

module Library
  Money = Data.define(:currency, :value)

  module_function

  def gbp(value)
    Money.new(currency: 'GBP', value: value)
  end

  FEE_PER_DAY = gbp(0.5)

  def add_money(a, b)
    raise ArgumentError, "cannot add #{b.currency} to #{a.currency}" if a.currency != b.currency

    Money.new(currency: a.currency, value: a.value + b.value)
  end

  def late_fee(loan, returned_on)
    days_late = [0, (returned_on - loan[:due]).to_i].max
    gbp(days_late * FEE_PER_DAY.value)
  end

  def may_borrow(loans, on)
    loans.all? { |loan| loan[:due] >= on }
  end
end

File: examples/csharp-vstest/src/Library.cs
namespace Varar.Example;

/// <summary>An amount of money in a single currency — mirrors the TypeScript sample's <c>Money</c>.</summary>
public sealed record Money(string Currency, decimal Value);

/// <summary>A borrowed book and the day it is due back.</summary>
public sealed record Loan(string Title, DateOnly Due);

// The code under test for library.md. Pure domain logic over DateOnly and Money:
// parsing and formatting of dates and money belong to the step definitions, not here.
public static class Library
{
    public static readonly Money FeePerDay = Gbp(0.50m);

    public static Money Gbp(decimal value) => new("GBP", value);

    public static Money AddMoney(Money a, Money b) =>
        a.Currency == b.Currency
            ? a with { Value = a.Value + b.Value }
            : throw new InvalidOperationException($"cannot add {b.Currency} to {a.Currency}");

    public static Money LateFee(Loan loan, DateOnly returnedOn) =>
        Gbp(Math.Max(0, returnedOn.DayNumber - loan.Due.DayNumber) * FeePerDay.Value);

    public static bool MayBorrow(IEnumerable<Loan> loans, DateOnly on) =>
        loans.All(loan => loan.Due >= on);
}
